### DISCUSSION TITLE
Write a custom middleware that gets executed quite early in the pipeline.

### DISCUSSION THREAD

**Discussion by mkgn** [OrchardCMS/OrchardCore#12995] [General]
My use case is that the CMS content is accessible by a limited set of clients. They will be sending a x-header with some encrypted data so that my middleware will decrypt and if it is OK, then only I should let it proceed to query any content. The Middleware module sample [here](https://github.com/OrchardCMS/OrchardCore.Samples) is self explanatory but I am not sure how to add my middleware quite early so that I don't want lot of other middleware to execute if that X-Header is not available/invalid.

Any ideas are appreciated.

**Comment by hishamco** [comment]
Your question is more ASP.NET Core specific, you can add your middleware before `app.UseOrchardCore()` get called

https://github.com/OrchardCMS/OrchardCore/blob/3ec70279b458af7dec53ee412717bb1d89f9f970/src/OrchardCore.Cms.Web/Program.cs#L20

**Reply by hishamco** [reply]
I think one of the best options is using the rate limiting middleware

**Comment by jtkech** [comment]
If the logic needs to be executed at the tenant level, for example by a module that can be enabled/disabled in a given tenant (a tenant is composed of modules and we have at least the Default tenant), can be done in your module startup.

        // Can be overridden to be executed earlier, 0 by default.
        public override int ConfigureOrder => -150;

        public override void Configure(IApplicationBuilder app, IEndpointRouteBuilder routes, IServiceProvider serviceProvider)
        {
            app.UseMiddleware<YourMiddleware>();
        }

Or from the web application (without using a module) but still at the tenant level.

    .AddOrchardCms()
    .AddSetupFeatures("OrchardCore.AutoSetup")
    .Configure((tenantAppBuilder) =>
    {
        tenantAppBuilder.UseMiddleware<YourMiddleware>();
    }, order: -150);

**Reply by hishamco** [reply]
This is another approach but `Order` is tricky coz you don't if there's an existing module has a less value ;)

**Reply by jtkech** [reply]
Yes I agree but this is another approach that you need if your middleware needs to inject tenant level services, and/or if you want to enable/disable it through a module feature in which it is defined, but maybe this is not the case here.

**Comment by mkgn** [comment]
I currently just wrote a test middleware and hooked it before app.UseOrchardCore(). It seems to do the trick. But I am interested in @hishamco suggestion as well. Going to explore it too. Thank you all for feedback.
